python/xir/program.py

```diff
@@ -637,6 +648,143 @@ def merge(*programs) -> "XIRProgram":
+
+        .. code-block:: python
+
+            # Define a library containing the XIR programs in the resolution scope.
+            library = {"main": main_program, "util": util_program}
+
+            # Resolve the imports from the main XIR program.
+            resolved_main_program = xir.XIRProgram.resolve(library, "main")
+
+            # Display the result.
+            print(resolved_main_program.serialize())
+
+        The output is
+
+        .. code-block:: none
+
+            gate H2, 0, 2;
+
+            gate H2:
+                H | [0];
+                H | [1];
+            end;
+
+            H2 | [0, 1];
+        """
+        resolved_names = XIRProgram._resolve_include_order(library, name, stack=set(), cache={})
+        resolved_programs = map(library.get, resolved_names)
+
+        resolved_program = XIRProgram.merge(*resolved_programs)
+        resolved_program._includes = []  # pylint: disable=protected-access
+        return resolved_program
+
+    @staticmethod
+    def _resolve_include_order(
+        library: Mapping[str, "XIRProgram"],
+        name: str,
+        stack: MutableSet[str],
+        cache: MutableMapping[str, Sequence[str]],
+    ) -> Sequence[str]:
+        """
+        Resolves the include order of an XIR program using C3 superclass linearization.
+
+        See https://en.wikipedia.org/wiki/C3_linearization for a summary of the
+        properties that hold following the linearization.
+
+        Args:
+            library (Mapping[str, XIRProgram]): mapping from names to XIR programs
+            name (str): name of the root XIR program to resolve
+            stack (MutableSet[str]): names in the current call stack; prevents
+                infinite recursion in the case of circular dependencies.
+            cache (MutableMapping[str, Sequence[str]]): mapping from names to
+                linearizations; improves performance when an XIR program is
+                included by multiple XIR programs.
+
+        Returns:
+            Sequence[str]: names representing a C3 linearization of the given XIR program.
+
+        Raises:
+            KeyError: if ``library`` does not have an entry for ``name``, or the
+                XIR program associated with ``name`` (transitively) includes
+                another XIR program which does not have an entry in ``library``.
+            ValueError: if ``name`` is already in ``stack``, or the XIR program
+                associated with ``name`` (transitively) includes an XIR program
+                with a circular dependency.
+        """
+        if name not in library:
+            raise KeyError(f"XIR program '{name}' cannot be found.")
+
+        if name in stack:
+            raise ValueError(f"XIR program '{name}' has a circular dependency.")
+
+        # The stack is only empty when processing the root XIR program.
+        if stack and library[name].statements:
+            raise ValueError(f"XIR program '{name}' contains a statement.")
```

Comment:
Oh, nice, so this basically just checks that there are no statements in included programs. Just to be a bit more explicit:
```suggestion
        if len(stack) !=0 and library[name].statements:
            raise ValueError(f"XIR program '{name}' contains a statement.")
```

Reply:
Actually, the current style is recommended by [PEP 8](https://www.python.org/dev/peps/pep-0008/) and [Stack Overflow](https://stackoverflow.com/a/53522).  Are you OK with leaving the code as-is?

Reply:
Yep, I'm completely fine with it. Good to follow PEPs! 😄